### DISCUSSION TITLE
How can I subscribe to a queue without using a decorator?

### DISCUSSION THREAD

**Discussion by YanNezamutdinov** [ag2ai/faststream#1411] [Q&A]
The microservice accepts a list of queue names. I create instances of the class, where I pass the name of the queue that I want to listen to.

I don't know exactly how to properly organize a connection to rabbitmq, inside a class, or transfer connect to an instance of the class, and I don't quite understand how to subscribe to a queue without a decorator?

Each microservice must subscribe to 10-100 queues.

**Comment by Lancetnik** [comment]
Sure, you can use router with explicit routes if you want: https://faststream.airt.ai/latest/getting-started/routers/#delay-handler-registration

**Reply by YanNezamutdinov** [reply]
Sorry, this example didn't help me. I only need to subscribe to the given topics.  Their number is arbitrary, from 1 to N. But not with the help of a decorator.

**Reply by Lancetnik** [reply]
```python
from faststream.kafka import KafkaRouter, KafkaRouter, KafkaBroker

def handler(msg):
    print(msg)
   
def setup_consumer(*topics: str) -> KafkaRouter:
    return KafkaRouter(handlers=[
        KafkaRoute(handler, *topics)
    ])
    
broker = KafkaBroker()
broker.include_router(setup_consumer("topic1", "topic2"))
```

**Reply by YanNezamutdinov** [reply]
Yes, it does work, but not when the application is already running. Judging by previous discussions, you need to call `broker.start()`, but `RabbitRouter` does not inherit this property from `RabbitBroker`. I really want a more explicit subscriber constructor. 

p.s. I really liked FastStream, thanks a lot for the work!

**Reply by Lancetnik** [reply]
Since 0.5.0 you are able to register subscribers dynamically without whole application restarting:

```python
subscriber = broker.subscriber("dynamic")
subscriber(handler_method)
...
broker.setup_subscriber(subscriber)
await subscriber.start()
...
await subscriber.close()
```
Is thit decision suitable for you?

P.S. Thanks for your feedback!

**Reply by Lancetnik** [reply]
Also, you can create subscriber by router, but it should be setted up via a broker object.

**Reply by citos88** [reply]
Hi, this method works for me, but I have a problem with typings in handler_method - as parameter I get  RabbitMessage (I use RabbitMQ), not pydantic object. Just to add that I use fastapi

**Reply by Lancetnik** [reply]
@citos88 sorry, it is not working with FastAPI yet

**Reply by YanNezamutdinov** [reply]
I tried it like this:

```python
from faststream.rabbit import RabbitExchange, RabbitQueue, RabbitBroker

async def handle_message(message):
    data = json.loads(message)
    print(data)

async def setup_consumer(queue: str):
    broker = RabbitBroker()
    subscriber = broker.subscriber(queue=RabbitQueue(queue),
                                   exchange=RabbitExchange("test_ext.direct"))
    subscriber(handle_message)
    broker.setup_subscriber(subscriber)
    await subscriber.start()
```
But I get an error `faststream.exceptions.SetupError: You should setup subscriber at first.`

**Reply by Lancetnik** [reply]
Broker should already connected to be able to setup subscriber. I have to add a new error message. Anyway, you are trying to make something strange and the framework was not designed for such usage. What's your case?

**Reply by YanNezamutdinov** [reply]
This is a microservice project. One of the services is waiting for a message with a set of queues to which it needs to connect and process the messages that arrive there. Sometimes you need to change the set of listening queues.

**Reply by Lancetnik** [reply]
Well, you can take a look here: this is the correct usage of dynamic subscibers: https://github.com/airtai/fastagency/blob/main/examples/nats/faststream_backend.py#L50

P.S. you should use already running broker to setup them, but register via router or any other broker

**Reply by YanNezamutdinov** [reply]
Yes, that's about what I did. But it's still difficult for me to understand the idea of inheriting brokers, probably because I'm new here. Is it a good idea to declare a broker `broker = NatsBroker("nats://nats-fastagency:4222")` outside of a function every time it is needed?

**Reply by Lancetnik** [reply]
Nope, broker should be the single instance in your app. It is important to use already running broker for all operations. You have all possibilities to do any you want without a new broker creation

**Reply by hasansezertasan** [reply]
Hey there, I have some questions. I implemented these two:

The first one is based on the [documentation](https://github.com/ag2ai/faststream/blob/1a2bc1dedb5cdbf069d75e89e616892c7e6a4144/docs/docs_src/integrations/http_frameworks_integrations/litestar.py#L4):

```python
# /// script
# requires-python = ">=3.12"
# dependencies = ["uvicorn", "litestar", "faststream[rabbit]"]
# ///
"""Litestar application with FastStream Integration."""
import uvicorn
from faststream.rabbit import RabbitBroker
from litestar import Litestar, get

# Define the broker
broker = RabbitBroker("amqp://guest:guest@localhost:5672/")


# Register the handler
@broker.subscriber("test-queue")
def handle(msg: str) -> None:
    print(msg)


# Define the endpoint
@get("/{name:str}")
async def index(name: str) -> str:
    message = f"Hello, {name}!"
    await broker.publish(message=message, queue="test-queue")
    return message


# Assamble the app
app = Litestar(
    [index],
    on_startup=[broker.start],
    on_shutdown=[broker.stop],
)

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8080)

```

This approach disturbs me from time to time, and in the end I somehow find myself trying to fix a circular-import exception. I've been in similar situations using Flask and FastAPI. I find the [Litestar approach](https://docs.litestar.dev/2/migration/fastapi.html) better. Then I decided to learn more about FastStream but couldn't find a way to implement register some function to the broker, until I found this discussion and I wrote this example:

```python
# /// script
# requires-python = ">=3.12"
# dependencies = ["uvicorn", "litestar", "faststream[rabbit]"]
# ///
"""Litestar application with FastStream Integration."""
import uvicorn
from faststream.rabbit import RabbitBroker, RabbitRoute, RabbitRouter
from litestar import Litestar, get


# Define the handler
def handle(msg: str) -> None:
    print(msg)


# Define the endpoint
@get("/{name:str}")
async def index(name: str) -> str:
    message = f"Hello, {name}!"
    await broker.publish(message=message, queue="test-queue")
    return message


# Define the broker
broker = RabbitBroker("amqp://guest:guest@localhost:5672/")

# Include the router
broker.include_router(RabbitRouter(handlers=[RabbitRoute(handle, queue="test-queue")]))


# Assamble the app
app = Litestar(
    [index],
    on_startup=[broker.start],
    on_shutdown=[broker.stop],
)

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8080)

```

Here are my questions:

- Is there any difference (internally) between these implementations? Do they work just the same?
- Is is possible to include the router to the broker when initializing it, like: 

```python
# /// script
# requires-python = ">=3.12"
# dependencies = ["uvicorn", "litestar", "faststream[rabbit]"]
# ///
"""Litestar application with FastStream Integration."""
import uvicorn
from faststream.rabbit import RabbitBroker, RabbitRoute, RabbitRouter
from litestar import Litestar, get


# Define the handler
def handle(msg: str) -> None:
    print(msg)


# Define the endpoint
@get("/{name:str}")
async def index(name: str) -> str:
    message = f"Hello, {name}!"
    await broker.publish(message=message, queue="test-queue")
    return message


# Assamble the app
broker = RabbitBroker(
    "amqp://guest:guest@localhost:5672/",
    routers=[RabbitRouter(handlers=[RabbitRoute(handle, queue="test-queue")])],
)


# Assamble the app
app = Litestar(
    [index],
    on_startup=[broker.start],
    on_shutdown=[broker.stop],
)

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8080)

```

---

Decorators that are bound to a singleton class and register some function or component in the bound instance... It's evil 👿! I have one side-note, adding a section to the documentation about bigger applications may be really helpful, something like:

- [Bigger Applications - FastAPI](https://fastapi.tiangolo.com/tutorial/bigger-applications/)
- [Bigger Applications - Rocketry](https://rocketry.readthedocs.io/en/stable/cookbook/bigger_applications.html)

**Reply by Lancetnik** [reply]
@hasansezertasan the recommended design is to register subscribers by router decorators and include routers in the `main.py` entrypoint of the application.

```python
# handler1.py
router = Router()

@router.subscriber(...)
async def handler1(): ...

# handler2.py
router = Router()

@router.subscriber(...)
async def handler2(): ...

# main.py
from handler1 import router as router1
from handler2 import router as router2

broker = Broker(routers=[router1, router2])
app = FastStream(broker)
```

`RabbitRouter(handlers=[RabbitRoute(handle, queue="test-queue")])` is an ugly solution I plan to deprecate after this feature implementation - https://github.com/ag2ai/faststream/issues/2451
